File: kitagawa-detect/weed_detector.py
"""
Weed detector using OpenCV.

Supported workflows:
- Detect weed regions in still images from images/source.
- Read camera settings from a config file and print the largest weed position
  in live camera frames.
"""

import argparse
import csv
import sys
from datetime import datetime
from pathlib import Path

import cv2
from weed_model import (
    HSV_LOWER,
    HSV_UPPER,
    LABEL_MAP,
    MIN_CONTOUR_AREA,
    MODEL_PATH,
    OUTPUT_DIR,
    SCALER_PATH,
    SOURCE_DIR,
    WeedClassifier,
    extract_features,
    get_image_paths,
)


def load_classifier():
    """Load the saved classifier."""
    if not MODEL_PATH.exists():
        print(f"[ERROR] Model file not found: {MODEL_PATH}")
        print("Run python learn.py first.")
        sys.exit(1)
    if not SCALER_PATH.exists():
        print(f"[ERROR] Scaler file not found: {SCALER_PATH}")
        print("Run python learn.py first.")
        sys.exit(1)

    print(f"Loading model: {MODEL_PATH}")
    print(f"Loading scaler: {SCALER_PATH}")
    try:
        return WeedClassifier().load(MODEL_PATH, SCALER_PATH)
    except RuntimeError as exc:
        print(f"[ERROR] {exc}")
        sys.exit(1)


def _contour_center(contour, bbox):
    """Return a contour center in image coordinates."""
    moments = cv2.moments(contour)
    if moments["m00"] != 0:
        center_x = int(moments["m10"] / moments["m00"])
        center_y = int(moments["m01"] / moments["m00"])
        return (center_x, center_y)

    x, y, w, h = bbox
    return (x + w // 2, y + h // 2)


def detect_weed_regions(image_bgr):
    """
    Detect weed candidate regions with HSV segmentation.

    Returns a list sorted by area descending.
    """
    hsv = cv2.cvtColor(image_bgr, cv2.COLOR_BGR2HSV)
    mask = cv2.inRange(hsv, HSV_LOWER, HSV_UPPER)

    kernel = cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (7, 7))
    mask = cv2.morphologyEx(mask, cv2.MORPH_CLOSE, kernel, iterations=2)
    mask = cv2.morphologyEx(mask, cv2.MORPH_OPEN, kernel, iterations=1)

    contours, _ = cv2.findContours(mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)

    img_h, img_w = image_bgr.shape[:2]
    total_area = img_h * img_w

    regions = []
    for contour in contours:
        area = cv2.contourArea(contour)
        if area < MIN_CONTOUR_AREA:
            continue

        bbox = cv2.boundingRect(contour)
        regions.append(
            {
                "bbox": bbox,
                "area_px": int(area),
                "area_ratio": round(area / total_area * 100, 2),
                "center": _contour_center(contour, bbox),
                "contour": contour,
            }
        )

    regions.sort(key=lambda region: region["area_px"], reverse=True)
    return regions


def annotate_image(
    image_bgr,
    regions,
    classification_label="live",
    classification_score=None,
):
    """Draw detection results on an image."""
    annotated = image_bgr.copy()
    img_h, img_w = annotated.shape[:2]

    if classification_score is None:
        header_text = classification_label
    else:
        header_text = f"Classification: {classification_label} ({classification_score:.1%})"

    cv2.rectangle(annotated, (0, 0), (img_w, 40), (0, 0, 0), -1)
    cv2.putText(
        annotated,
        header_text,
        (10, 28),
        cv2.FONT_HERSHEY_SIMPLEX,
        0.8,
        (0, 255, 255),
        2,
    )

    total_weed_area = 0
    for index, region in enumerate(regions, start=1):
        x, y, w, h = region["bbox"]
        center_x, center_y = region["center"]
        total_weed_area += region["area_px"]

        cv2.rectangle(annotated, (x, y), (x + w, y + h), (0, 255, 0), 2)

        overlay = annotated.copy()
        cv2.drawContours(overlay, [region["contour"]], -1, (0, 200, 0), -1)
        cv2.addWeighted(overlay, 0.3, annotated, 0.7, 0, annotated)

        cv2.circle(annotated, (center_x, center_y), 5, (0, 0, 255), -1)

        label = (
            f"#{index} center=({center_x},{center_y}) "
            f"bbox={w}x{h}px area={region['area_px']}px ({region['area_ratio']}%)"
        )
        label_y = max(y - 10, 50)
        (text_w, text_h), _ = cv2.getTextSize(
            label, cv2.FONT_HERSHEY_SIMPLEX, 0.5, 1
        )
        cv2.rectangle(
            annotated,
            (x, label_y - text_h - 4),
            (x + text_w + 4, label_y + 4),
            (0, 0, 0),
            -1,
        )
        cv2.putText(
            annotated,
            label,
            (x + 2, label_y),
            cv2.FONT_HERSHEY_SIMPLEX,
            0.5,
            (0, 255, 0),
            1,
        )

    total_ratio = round(total_weed_area / (img_h * img_w) * 100, 2) if total_weed_area else 0
    footer = f"Weed regions: {len(regions)} | Total area: {total_weed_area}px ({total_ratio}%)"
    cv2.rectangle(annotated, (0, img_h - 35), (img_w, img_h), (0, 0, 0), -1)
    cv2.putText(
        annotated,
        footer,
        (10, img_h - 10),
        cv2.FONT_HERSHEY_SIMPLEX,
        0.7,
        (0, 255, 255),
        2,
    )

    return annotated


def process_source_images():
    """Run detection for all still images in images/source."""
    print("=" * 60)
    print("Processing source images")
    print("=" * 60)

    source_files = get_image_paths(SOURCE_DIR)
    if not source_files:
        print(f"[ERROR] No source images found in: {SOURCE_DIR}")
        sys.exit(1)

    print(f"Source images: {len(source_files)}")

    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
    output_csv_path = OUTPUT_DIR / "detections.csv"

    classifier = load_classifier()
    csv_rows = []

    for index, img_path in enumerate(source_files, start=1):
        print(f"[{index}/{len(source_files)}] Processing {img_path.name} ...")
        image_bgr = cv2.imread(str(img_path))
        if image_bgr is None:
            print("  -> failed to read image")
            continue

        feat = extract_features(image_bgr)
        pred, score = classifier.predict_with_confidence(feat)
        label = LABEL_MAP.get(pred, "unknown")

        print(f"  Classification: {label} (confidence: {score:.2%})")

        regions = detect_weed_regions(image_bgr)
        print(f"  Weed regions: {len(regions)}")

        if regions:
            largest = regions[0]
            print(
                "  Largest weed center: "
                f"({largest['center'][0]}, {largest['center'][1]}) px"
            )
            csv_rows.append(
                {
                    "timestamp": datetime.now().astimezone().isoformat(timespec="seconds"),
                    "x": largest["center"][0],
                    "y": largest["center"][1],
                    "size": largest["area_px"],
                }
            )
        else:
            csv_rows.append(
                {
                    "timestamp": datetime.now().astimezone().isoformat(timespec="seconds"),
                    "x": "",
                    "y": "",
                    "size": 0,
                }
            )

        for region_index, region in enumerate(regions, start=1):
            x, y, w, h = region["bbox"]
            center_x, center_y = region["center"]
            print(
                f"    #{region_index}: center=({center_x}, {center_y}) px, "
                f"bbox=({x}, {y}, {w}, {h}), "
                f"area={region['area_px']}px ({region['area_ratio']}%)"
            )

        print()

    with output_csv_path.open("w", newline="", encoding="utf-8") as csv_file:
        writer = csv.DictWriter(csv_file, fieldnames=["timestamp", "x", "y", "size"])
        writer.writeheader()
        writer.writerows(csv_rows)

    print("=" * 60)
    print(f"Done. Detection CSV was saved to: {output_csv_path}")
    print("=" * 60)


def _parse_camera_source(raw_value):
    """Convert a numeric camera source to int, keep device paths as strings."""
    value = raw_value.strip()
    if value.lstrip("-").isdigit():
        return int(value)
    return value


def load_camera_config(config_path):
    """Load camera settings from a key=value config file."""
    path = Path(config_path)
    if not path.exists():
        print(f"[ERROR] Config file not found: {path}")
        sys.exit(1)

    config = {}
    try:
        with path.open("r", encoding="utf-8") as config_file:
            for line_number, raw_line in enumerate(config_file, start=1):
                line = raw_line.strip()
                if not line or line.startswith("#"):
                    continue
                if "=" not in line:
                    raise ValueError(
                        f"Invalid config line {line_number}: {raw_line.rstrip()}"
                    )
                key, value = line.split("=", 1)
                key = key.strip()
                value = value.strip()
                if not key:
                    raise ValueError(f"Empty key at line {line_number}")
                config[key] = value
    except ValueError as exc:
        print(f"[ERROR] {exc}")
        sys.exit(1)

    required_keys = ("camera", "camera_width", "camera_height")
    missing_keys = [key for key in required_keys if key not in config]
    if missing_keys:
        print(f"[ERROR] Missing config keys: {', '.join(missing_keys)}")
        sys.exit(1)

    try:
        camera_width = int(config["camera_width"])
        camera_height = int(config["camera_height"])
    except ValueError:
        print("[ERROR] camera_width and camera_height must be integers.")
        sys.exit(1)

    if camera_width <= 0 or camera_height <= 0:
        print("[ERROR] camera_width and camera_height must be positive.")
        sys.exit(1)

    return {
        "config_path": path,
        "camera_raw": config["camera"],
        "camera": _parse_camera_source(config["camera"]),
        "camera_width": camera_width,
        "camera_height": camera_height,
    }


def open_camera(camera_config):
    """Open the configured camera and apply the requested frame size."""
    capture = cv2.VideoCapture(camera_config["camera"])
    if not capture.isOpened():
        print(f"[ERROR] Failed to open camera: {camera_config['camera_raw']}")
        sys.exit(1)

    capture.set(cv2.CAP_PROP_FRAME_WIDTH, camera_config["camera_width"])
    capture.set(cv2.CAP_PROP_FRAME_HEIGHT, camera_config["camera_height"])
    return capture


def _print_largest_weed_console(region):
    """Continuously update the console with the largest weed position."""
    if region is None:
        print("\rLargest weed center: not detected", end="", flush=True)
        return

    center_x, center_y = region["center"]
    x, y, w, h = region["bbox"]
    print(
        "\r"
        f"Largest weed center: ({center_x}, {center_y}) px | "
        f"bbox=({x}, {y}, {w}, {h}) | area={region['area_px']}px      ",
        end="",
        flush=True,
    )


def process_camera_stream(config_path):
    """Read frames from the configured camera and print the largest weed center."""
    camera_config = load_camera_config(config_path)

    print("=" * 60)
    print("Starting live camera detection")
    print("=" * 60)
    print(f"Config: {camera_config['config_path']}")
    print(f"Camera: {camera_config['camera_raw']}")
    print(
        "Requested size: "
        f"{camera_config['camera_width']}x{camera_config['camera_height']}"
    )

    capture = open_camera(camera_config)
    actual_width = int(capture.get(cv2.CAP_PROP_FRAME_WIDTH))
    actual_height = int(capture.get(cv2.CAP_PROP_FRAME_HEIGHT))
    if actual_width > 0 and actual_height > 0:
        print(f"Opened size: {actual_width}x{actual_height}")

    print("Press Ctrl+C to stop.")

    try:
        while True:
            ok, frame = capture.read()
            if not ok or frame is None:
                print("\n[ERROR] Failed to read a frame from the camera.")
                break

            regions = detect_weed_regions(frame)
            largest = regions[0] if regions else None
            _print_largest_weed_console(largest)
    except KeyboardInterrupt:
        print("\nStopped live camera detection.")
    finally:
        capture.release()
        print()


def build_arg_parser():
    """Build the command line parser."""
    parser = argparse.ArgumentParser(description="Weed detector")
    parser.add_argument(
        "--detect",
        action="store_true",
        help="process still images from images/source/",
    )
    parser.add_argument(
        "--camera-detect",
        action="store_true",
        help="read the camera from --config and print the largest weed center",
    )
    parser.add_argument(
        "--config",
        type=Path,
        help="path to a config.txt file with camera settings",
    )
    return parser


def main():
    """CLI entry point."""
    parser = build_arg_parser()
    args = parser.parse_args()

    if not any([args.detect, args.camera_detect]):
        print("Examples:")
        print("  python weed_detector.py --detect")
        print("  python weed_detector.py --camera-detect --config config.txt")
        print()
        parser.print_help()
        sys.exit(0)

    if args.detect:
        process_source_images()

    if args.camera_detect:
        if args.config is None:
            parser.error("--camera-detect requires --config PATH")
        process_camera_stream(args.config)


if __name__ == "__main__":
    main()

File: kitagawa-detect/weed_detector.cpp
#include <algorithm>
#include <cctype>
#include <chrono>
#include <cmath>
#include <cstdlib>
#include <ctime>
#include <filesystem>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <optional>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

#include <opencv2/core.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/ml.hpp>
#include <opencv2/videoio.hpp>

namespace fs = std::filesystem;

namespace {

const fs::path kBaseDir = fs::current_path();
const fs::path kSourceDir = kBaseDir / "images" / "source";
const fs::path kOutputDir = kBaseDir / "output";
const fs::path kModelDir = kBaseDir / "model";
const fs::path kModelPath = kModelDir / "weed_classifier.yml";
const fs::path kScalerPath = kModelDir / "weed_scaler.yml";

const cv::Size kImageSize(128, 128);
const cv::Scalar kHsvLower(25, 40, 40);
const cv::Scalar kHsvUpper(95, 255, 255);
constexpr double kMinContourArea = 500.0;

const std::vector<std::string> kImageExtensions = {
    ".jpg",
    ".jpeg",
    ".png",
    ".bmp",
    ".webp",
};

struct WeedRegion {
    cv::Rect bbox;
    int areaPx = 0;
    double areaRatio = 0.0;
    cv::Point center;
    std::vector<cv::Point> contour;
};

struct CameraConfig {
    fs::path configPath;
    std::string cameraRaw;
    bool useIndex = false;
    int cameraIndex = 0;
    int cameraWidth = 0;
    int cameraHeight = 0;
};

struct CsvRow {
    std::string timestamp;
    std::string x;
    std::string y;
    int size = 0;
};

class WeedClassifier {
public:
    void load(const fs::path& modelPath, const fs::path& scalerPath) {
        if (!fs::exists(modelPath)) {
            throw std::runtime_error("model file not found: " + modelPath.string());
        }
        if (!fs::exists(scalerPath)) {
            throw std::runtime_error("scaler file not found: " + scalerPath.string());
        }

        svm_ = cv::Algorithm::load<cv::ml::SVM>(modelPath.string());
        if (svm_.empty()) {
            throw std::runtime_error("failed to load model: " + modelPath.string());
        }

        cv::FileStorage fsStorage(scalerPath.string(), cv::FileStorage::READ);
        if (!fsStorage.isOpened()) {
            throw std::runtime_error("failed to open scaler file for reading");
        }
        fsStorage["feature_mean"] >> featureMean_;
        fsStorage["feature_std"] >> featureStd_;

        if (featureMean_.empty() || featureStd_.empty()) {
            throw std::runtime_error("invalid scaler data");
        }
    }

    int predict(const cv::Mat& sample, double* confidence = nullptr) const {
        if (svm_.empty()) {
            throw std::runtime_error("classifier is not loaded");
        }

        cv::Mat normalized = normalize(sample);
        float predicted = svm_->predict(normalized);

        if (confidence != nullptr) {
            cv::Mat rawOutput;
            svm_->predict(normalized, rawOutput, cv::ml::StatModel::RAW_OUTPUT);
            double margin = std::abs(rawOutput.at<float>(0, 0));
            *confidence = 1.0 / (1.0 + std::exp(-margin));
        }

        return static_cast<int>(predicted);
    }

private:
    cv::Mat normalize(const cv::Mat& samples) const {
        cv::Mat repeatedMean;
        cv::repeat(featureMean_, samples.rows, 1, repeatedMean);
        cv::Mat repeatedStd;
        cv::repeat(featureStd_, samples.rows, 1, repeatedStd);

        cv::Mat normalized;
        cv::subtract(samples, repeatedMean, normalized);
        cv::divide(normalized, repeatedStd, normalized);
        return normalized;
    }

    cv::Mat featureMean_;
    cv::Mat featureStd_;
    cv::Ptr<cv::ml::SVM> svm_;
};

std::string toLower(std::string value) {
    std::transform(
        value.begin(),
        value.end(),
        value.begin(),
        [](unsigned char ch) { return static_cast<char>(std::tolower(ch)); }
    );
    return value;
}

std::string trim(const std::string& value) {
    const auto begin = value.find_first_not_of(" \t\r\n");
    if (begin == std::string::npos) {
        return "";
    }
    const auto end = value.find_last_not_of(" \t\r\n");
    return value.substr(begin, end - begin + 1);
}

std::string nowTimestamp() {
    const auto now = std::chrono::system_clock::now();
    const std::time_t nowTime = std::chrono::system_clock::to_time_t(now);

    std::tm localTm{};
#ifdef _WIN32
    localtime_s(&localTm, &nowTime);
#else
    localtime_r(&nowTime, &localTm);
#endif

    std::ostringstream oss;
    oss << std::put_time(&localTm, "%Y-%m-%dT%H:%M:%S");
    return oss.str();
}

std::vector<fs::path> getImagePaths(const fs::path& directory) {
    std::vector<fs::path> paths;
    if (!fs::exists(directory) || !fs::is_directory(directory)) {
        return paths;
    }

    for (const auto& entry : fs::directory_iterator(directory)) {
        if (!entry.is_regular_file()) {
            continue;
        }

        const std::string extension = toLower(entry.path().extension().string());
        if (std::find(kImageExtensions.begin(), kImageExtensions.end(), extension) !=
            kImageExtensions.end()) {
            paths.push_back(entry.path());
        }
    }

    std::sort(paths.begin(), paths.end());
    return paths;
}

cv::Mat extractFeatures(const cv::Mat& imageBgr) {
    cv::Mat resized;
    cv::resize(imageBgr, resized, kImageSize);

    cv::Mat hsv;
    cv::cvtColor(resized, hsv, cv::COLOR_BGR2HSV);

    std::vector<float> features;
    features.reserve(101);

    for (int channel = 0; channel < 3; ++channel) {
        const int histSize[] = {32};
        const float range[] = {0.0f, 256.0f};
        const float* ranges[] = {range};
        const int channels[] = {channel};
        cv::Mat hist;
        cv::calcHist(&hsv, 1, channels, cv::Mat(), hist, 1, histSize, ranges);
        cv::normalize(hist, hist);

        for (int i = 0; i < hist.rows; ++i) {
            features.push_back(hist.at<float>(i, 0));
        }
    }

    cv::Mat greenMask;
    cv::inRange(hsv, kHsvLower, kHsvUpper, greenMask);
    const double greenRatio =
        static_cast<double>(cv::countNonZero(greenMask)) / static_cast<double>(greenMask.total());
    features.push_back(static_cast<float>(greenRatio));

    cv::Mat gray;
    cv::cvtColor(resized, gray, cv::COLOR_BGR2GRAY);

    cv::Mat sobelX;
    cv::Mat sobelY;
    cv::Sobel(gray, sobelX, CV_64F, 1, 0, 3);
    cv::Sobel(gray, sobelY, CV_64F, 0, 1, 3);

    cv::Mat magnitude;
    cv::magnitude(sobelX, sobelY, magnitude);

    cv::Scalar mean;
    cv::Scalar stddev;
    cv::meanStdDev(magnitude, mean, stddev);
    double maxValue = 0.0;
    cv::minMaxLoc(magnitude, nullptr, &maxValue);
    cv::Mat thresholdMask = magnitude > mean[0];
    const double activeRatio =
        static_cast<double>(cv::countNonZero(thresholdMask)) / static_cast<double>(thresholdMask.total());

    features.push_back(static_cast<float>(mean[0]));
    features.push_back(static_cast<float>(stddev[0]));
    features.push_back(static_cast<float>(maxValue));
    features.push_back(static_cast<float>(activeRatio));

    cv::Mat featureRow(1, static_cast<int>(features.size()), CV_32F);
    for (int i = 0; i < featureRow.cols; ++i) {
        featureRow.at<float>(0, i) = features[static_cast<size_t>(i)];
    }
    return featureRow;
}

cv::Point contourCenter(const std::vector<cv::Point>& contour, const cv::Rect& bbox) {
    const cv::Moments moments = cv::moments(contour);
    if (std::abs(moments.m00) > 1e-6) {
        return cv::Point(
            static_cast<int>(moments.m10 / moments.m00),
            static_cast<int>(moments.m01 / moments.m00)
        );
    }

    return cv::Point(bbox.x + bbox.width / 2, bbox.y + bbox.height / 2);
}

std::vector<WeedRegion> detectWeedRegions(const cv::Mat& imageBgr) {
    cv::Mat hsv;
    cv::cvtColor(imageBgr, hsv, cv::COLOR_BGR2HSV);

    cv::Mat mask;
    cv::inRange(hsv, kHsvLower, kHsvUpper, mask);

    const cv::Mat kernel =
        cv::getStructuringElement(cv::MORPH_ELLIPSE, cv::Size(7, 7));
    cv::morphologyEx(mask, mask, cv::MORPH_CLOSE, kernel, cv::Point(-1, -1), 2);
    cv::morphologyEx(mask, mask, cv::MORPH_OPEN, kernel, cv::Point(-1, -1), 1);

    std::vector<std::vector<cv::Point>> contours;
    cv::findContours(mask, contours, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_SIMPLE);

    const double totalArea = static_cast<double>(imageBgr.rows * imageBgr.cols);
    std::vector<WeedRegion> regions;

    for (const auto& contour : contours) {
        const double area = cv::contourArea(contour);
        if (area < kMinContourArea) {
            continue;
        }

        const cv::Rect bbox = cv::boundingRect(contour);
        WeedRegion region;
        region.bbox = bbox;
        region.areaPx = static_cast<int>(area);
        region.areaRatio = (area / totalArea) * 100.0;
        region.center = contourCenter(contour, bbox);
        region.contour = contour;
        regions.push_back(region);
    }

    std::sort(
        regions.begin(),
        regions.end(),
        [](const WeedRegion& lhs, const WeedRegion& rhs) { return lhs.areaPx > rhs.areaPx; }
    );
    return regions;
}

CameraConfig loadCameraConfig(const fs::path& configPath) {
    if (!fs::exists(configPath)) {
        throw std::runtime_error("config file not found: " + configPath.string());
    }

    std::ifstream input(configPath);
    if (!input.is_open()) {
        throw std::runtime_error("failed to open config file: " + configPath.string());
    }

    std::string cameraRaw;
    std::optional<int> width;
    std::optional<int> height;

    std::string line;
    int lineNumber = 0;
    while (std::getline(input, line)) {
        ++lineNumber;
        const std::string trimmed = trim(line);
        if (trimmed.empty() || trimmed.front() == '#') {
            continue;
        }

        const auto delimiter = trimmed.find('=');
        if (delimiter == std::string::npos) {
            throw std::runtime_error(
                "invalid config line " + std::to_string(lineNumber) + ": " + trimmed
            );
        }

        const std::string key = trim(trimmed.substr(0, delimiter));
        const std::string value = trim(trimmed.substr(delimiter + 1));
        if (key.empty()) {
            throw std::runtime_error("empty config key at line " + std::to_string(lineNumber));
        }

        if (key == "camera") {
            cameraRaw = value;
        } else if (key == "camera_width") {
            width = std::stoi(value);
        } else if (key == "camera_height") {
            height = std::stoi(value);
        }
    }

    if (cameraRaw.empty() || !width.has_value() || !height.has_value()) {
        throw std::runtime_error("config requires camera, camera_width, camera_height");
    }
    if (*width <= 0 || *height <= 0) {
        throw std::runtime_error("camera_width and camera_height must be positive");
    }

    CameraConfig config;
    config.configPath = configPath;
    config.cameraRaw = cameraRaw;
    config.cameraWidth = *width;
    config.cameraHeight = *height;

    const bool numericSource = !cameraRaw.empty() &&
        std::all_of(cameraRaw.begin(), cameraRaw.end(), [](unsigned char ch) { return std::isdigit(ch) != 0; });

    if (numericSource) {
        config.useIndex = true;
        config.cameraIndex = std::stoi(cameraRaw);
    }

    return config;
}

cv::VideoCapture openCamera(const CameraConfig& config) {
    cv::VideoCapture capture;
    if (config.useIndex) {
        capture.open(config.cameraIndex);
    } else {
        capture.open(config.cameraRaw);
    }

    if (!capture.isOpened()) {
        throw std::runtime_error("failed to open camera: " + config.cameraRaw);
    }

    capture.set(cv::CAP_PROP_FRAME_WIDTH, config.cameraWidth);
    capture.set(cv::CAP_PROP_FRAME_HEIGHT, config.cameraHeight);
    return capture;
}

void writeDetectionCsv(const fs::path& csvPath, const std::vector<CsvRow>& rows) {
    std::ofstream output(csvPath);
    if (!output.is_open()) {
        throw std::runtime_error("failed to open csv for writing: " + csvPath.string());
    }

    output << "timestamp,x,y,size\n";
    for (const auto& row : rows) {
        output << row.timestamp << "," << row.x << "," << row.y << "," << row.size << "\n";
    }
}

void processSourceImages() {
    std::cout << "============================================================\n";
    std::cout << "Processing source images with C++ detector\n";
    std::cout << "============================================================\n";

    const auto sourceFiles = getImagePaths(kSourceDir);
    if (sourceFiles.empty()) {
        throw std::runtime_error("no source images found in " + kSourceDir.string());
    }

    WeedClassifier classifier;
    classifier.load(kModelPath, kScalerPath);

    fs::create_directories(kOutputDir);
    const fs::path csvPath = kOutputDir / "detections_cpp.csv";

    std::vector<CsvRow> rows;
    rows.reserve(sourceFiles.size());

    for (size_t index = 0; index < sourceFiles.size(); ++index) {
        const auto& path = sourceFiles[index];
        std::cout << "[" << (index + 1) << "/" << sourceFiles.size() << "] Processing "
                  << path.filename().string() << " ...\n";

        const cv::Mat image = cv::imread(path.string());
        if (image.empty()) {
            std::cout << "  -> failed to read image\n";
            continue;
        }

        double confidence = 0.0;
        const int predicted = classifier.predict(extractFeatures(image), &confidence);
        const std::string label = predicted == 1 ? "weed" : "not_weed";
        std::cout << "  Classification: " << label
                  << " (confidence: " << std::fixed << std::setprecision(2)
                  << confidence * 100.0 << "%)\n";

        const auto regions = detectWeedRegions(image);
        std::cout << "  Weed regions: " << regions.size() << "\n";

        CsvRow row;
        row.timestamp = nowTimestamp();

        if (!regions.empty()) {
            const auto& largest = regions.front();
            row.x = std::to_string(largest.center.x);
            row.y = std::to_string(largest.center.y);
            row.size = largest.areaPx;

            std::cout << "  Largest weed center: (" << largest.center.x
                      << ", " << largest.center.y << ") px\n";
        } else {
            row.size = 0;
        }
        rows.push_back(row);

        for (size_t regionIndex = 0; regionIndex < regions.size(); ++regionIndex) {
            const auto& region = regions[regionIndex];
            std::cout << "    #" << (regionIndex + 1)
                      << ": center=(" << region.center.x << ", " << region.center.y << ") px, "
                      << "bbox=(" << region.bbox.x << ", " << region.bbox.y << ", "
                      << region.bbox.width << ", " << region.bbox.height << "), "
                      << "area=" << region.areaPx << "px ("
                      << std::fixed << std::setprecision(2) << region.areaRatio << "%)\n";
        }

        std::cout << "\n";
    }

    writeDetectionCsv(csvPath, rows);

    std::cout << "============================================================\n";
    std::cout << "Done. Detection CSV was saved to: " << csvPath << "\n";
    std::cout << "============================================================\n";
}

void processCameraStream(const fs::path& configPath) {
    const CameraConfig cameraConfig = loadCameraConfig(configPath);

    std::cout << "============================================================\n";
    std::cout << "Starting live camera detection (C++)\n";
    std::cout << "============================================================\n";
    std::cout << "Config: " << cameraConfig.configPath << "\n";
    std::cout << "Camera: " << cameraConfig.cameraRaw << "\n";
    std::cout << "Requested size: " << cameraConfig.cameraWidth
              << "x" << cameraConfig.cameraHeight << "\n";

    cv::VideoCapture capture = openCamera(cameraConfig);
    const int actualWidth = static_cast<int>(capture.get(cv::CAP_PROP_FRAME_WIDTH));
    const int actualHeight = static_cast<int>(capture.get(cv::CAP_PROP_FRAME_HEIGHT));
    if (actualWidth > 0 && actualHeight > 0) {
        std::cout << "Opened size: " << actualWidth << "x" << actualHeight << "\n";
    }

    std::cout << "Press Ctrl+C to stop.\n";

    while (true) {
        cv::Mat frame;
        if (!capture.read(frame) || frame.empty()) {
            throw std::runtime_error("failed to read a frame from the camera");
        }

        const auto regions = detectWeedRegions(frame);
        if (regions.empty()) {
            std::cout << "\rLargest weed center: not detected      " << std::flush;
            continue;
        }

        const auto& largest = regions.front();
        std::cout << "\rLargest weed center: (" << largest.center.x << ", " << largest.center.y
                  << ") px | bbox=(" << largest.bbox.x << ", " << largest.bbox.y << ", "
                  << largest.bbox.width << ", " << largest.bbox.height << ") | area="
                  << largest.areaPx << "px      " << std::flush;
    }
}

struct Options {
    bool detect = false;
    bool cameraDetect = false;
    std::optional<fs::path> configPath;
};

void printUsage(const char* programName) {
    std::cout << "Usage:\n";
    std::cout << "  " << programName << " --detect\n";
    std::cout << "  " << programName << " --camera-detect --config config.txt\n";
}

Options parseArgs(int argc, char** argv) {
    Options options;

    for (int i = 1; i < argc; ++i) {
        const std::string arg = argv[i];
        if (arg == "--detect") {
            options.detect = true;
        } else if (arg == "--camera-detect") {
            options.cameraDetect = true;
        } else if (arg == "--config") {
            if (i + 1 >= argc) {
                throw std::runtime_error("--config requires a path");
            }
            options.configPath = fs::path(argv[++i]);
        } else if (arg == "--help" || arg == "-h") {
            printUsage(argv[0]);
            std::exit(0);
        } else {
            throw std::runtime_error("unknown argument: " + arg);
        }
    }

    return options;
}

}  // namespace

int main(int argc, char** argv) {
    try {
        const Options options = parseArgs(argc, argv);

        if (!options.detect && !options.cameraDetect) {
            printUsage(argv[0]);
            return 0;
        }

        if (options.detect) {
            processSourceImages();
        }

        if (options.cameraDetect) {
            if (!options.configPath.has_value()) {
                throw std::runtime_error("--camera-detect requires --config PATH");
            }
            processCameraStream(*options.configPath);
        }

        return 0;
    } catch (const std::exception& ex) {
        std::cerr << "[ERROR] " << ex.what() << "\n";
        return 1;
    }
}
